eth/tracers/native/erc7562.go

```diff
@@ -0,0 +1,595 @@
+	From         common.Address  `json:"from"`
+	Gas          uint64          `json:"gas"`
+	GasUsed      uint64          `json:"gasUsed"`
+	To           *common.Address `json:"to,omitempty" rlp:"optional"`
+	Input        []byte          `json:"input" rlp:"optional"`
+	Output       []byte          `json:"output,omitempty" rlp:"optional"`
+	Error        string          `json:"error,omitempty" rlp:"optional"`
+	RevertReason string          `json:"revertReason,omitempty"`
+	//Calls        []callFrame     `json:"calls,omitempty" rlp:"optional"`
+	Logs []callLog `json:"logs,omitempty" rlp:"optional"`
+	// Placed at end on purpose. The RLP will be decoded to 0 instead of
+	// nil if there are non-empty elements after in the struct.
+	Value            *big.Int `json:"value,omitempty" rlp:"optional"`
+	revertedSnapshot bool
+
+	AccessedSlots     accessedSlots        `json:"accessedSlots"`
+	ExtCodeAccessInfo []common.Address     `json:"extCodeAccessInfo"`
+	DeployedContracts []common.Address     `json:"deployedContracts"`
+	UsedOpcodes       map[vm.OpCode]uint64 `json:"usedOpcodes"`
+	//GasObserved       bool                   `json:"gasObserved"`
+	ContractSize map[common.Address]*contractSizeWithOpcode `json:"contractSize"`
+	OutOfGas     bool                                       `json:"outOfGas"`
+	Calls        []callFrameWithOpcodes                     `json:"calls,omitempty" rlp:"optional"`
+}
+
+func (f callFrameWithOpcodes) TypeString() string {
+	return f.Type.String()
+}
+
+func (f callFrameWithOpcodes) failed() bool {
+	return len(f.Error) > 0 && f.revertedSnapshot
+}
+
+func (f *callFrameWithOpcodes) processOutput(output []byte, err error, reverted bool) {
+	output = common.CopyBytes(output)
+	// Clear error if tx wasn't reverted. This happened
+	// for pre-homestead contract storage OOG.
+	if err != nil && !reverted {
+		err = nil
+	}
+	if err == nil {
+		f.Output = output
+		return
+	}
+	f.Error = err.Error()
+	f.revertedSnapshot = reverted
+	if f.Type == vm.CREATE || f.Type == vm.CREATE2 {
+		f.To = nil
+	}
+	if !errors.Is(err, vm.ErrExecutionReverted) || len(output) == 0 {
+		return
+	}
+	f.Output = output
+	if len(output) < 4 {
+		return
+	}
+	if unpacked, err := abi.UnpackRevert(output); err == nil {
+		f.RevertReason = unpacked
+	}
+}
+
+type callFrameWithOpcodesMarshaling struct {
+	TypeString string `json:"type"`
+	Gas        hexutil.Uint64
+	GasUsed    hexutil.Uint64
+	Value      *hexutil.Big
+	Input      hexutil.Bytes
+	Output     hexutil.Bytes
+}
+
+// TODO: I suggest that we provide an `[]string` for all of these fields. Doing it for `Reads` as an example here.
+type accessedSlots struct {
+	Reads           map[string][]string `json:"reads"`
+	Writes          map[string]uint64   `json:"writes"`
+	TransientReads  map[string]uint64   `json:"transientReads"`
+	TransientWrites map[string]uint64   `json:"transientWrites"`
+}
+
+type opcodeWithPartialStack struct {
+	Opcode    vm.OpCode
+	StackTop3 []uint256.Int
```

Comment:
Let's not hard-code the number of opcodes kept, and instead have it a config parameter? I can totally see someday getting a call saying that we need to enable an opcode sequence of `DUP SWAP POP JUMP SQUAT SIT BARK` and the fate of the world depends on it.

Reply:
This is not opcodes kept but stack items for each opcode. We can configure that, it was 3 since we don't need more atm. Would we need more stack items in the future?